### PR TITLE
Fetch list of upgradable packages for fledge repository FOGL-6888.

### PR DESCRIPTION
**Pull request by deepanshu-yadav**
How to test 
1. Add deb http://archives.fledge-iot.org/latest/ubuntu1804/x86_64/ /  to /etc/apt/sources.list  Depending on your environment.
2. Install some packages like fledge , fledge-south-sinusoid 
3. Copy the full update.py inside `sudo vim /usr/local/fledge/python/fledge/services/core/api/update.py` 
4. However only add `app.router.add_route('GET', '/fledge/update', update.get_updates)` inside   `sudo vim /usr/local/fledge/python/fledge/services/core/routes.py`
5. Delete the line at step 1 and add deb http://archives.fledge-iot.org/nightly/ubuntu1804/x86_64/ / depending on your environment.  
6. Restart fledge.
7. Run `curl -sX GET http://localhost:8081/fledge/update  |jq`
8. If packages are available then response will be 
```
{
  "updates": [
    "fledge-south-sinusoid","fledge"
  ]
}
```
9. If no updates then 
```
{
  "updates": [
  ]
}
```

### CODE REVIEW

File: python/fledge/services/core/api/update.py [hunk 1]
```diff
@@ -107,3 +110,44 @@ async def update_package(request):
         raise web.HTTPInternalServerError(reason=msg, body=json.dumps({"message": msg}))
     else:
         return web.json_response({"status": "Running", "message": status_message})
+
+
+async def get_updates(request: web.Request) -> web.Response:
```
Comment (MarkRiddoch) [inline python/fledge/services/core/api/update.py:115]: Please add comment block as per normal practice

**Comment by deepanshu-yadav** [comment]
> @deepanshu-yadav In FogLAMP Mange there is an API that calls /foglamp/update for the manage instance
> 
> Please make sure this new change is still working in FogLAMP Mange

It is "GET" not PUT.

**Comment by deepanshu-yadav** [comment]
Hi @praveen-garg , @MarkRiddoch , @pintomax , 

I am getting the following output when i run the following command. 
```
ubuntu@ip-10-0-0-180:~$ apt list --upgradable | grep \^fledge

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

fledge-gui/unknown 1.9.2-440-gf849eed5 all [upgradable from: 1.9.2]
fledge-south-sinusoid/unknown 1.9.2-1-g38a138f amd64 [upgradable from: 1.9.2]
ubuntu@ip-10-0-0
```

Infact the API is also mimicking the same behavior.  

```
ubuntu@ip-10-0-0-180:~$ curl -sX GET http://localhost:8081/fledge/update |jq
{
  "updates": [
    "fledge-gui",
    "fledge-south-sinusoid"
  ]
}
```
When the user will upgrade these packages then isn't fledge is supposed to be there. 
Please guide.

**Comment by praveen-garg** [comment]
> When the user will upgrade these packages then isn't fledge is supposed to be there. Please guide.

There might be no new revision, ensure checking cmd run from latest to 2.0.0/nightly with a python sandbox code.

also list apt list upgradable without grep.

File: python/fledge/services/core/api/update.py [hunk 2]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+            ]
+        }
+        Example
+         curl -sX GET http://localhost:8081/fledge/update |jq
+    """
+    _platform = platform.platform()
+    if "centos" in _platform or "redhat" in _platform:
+        update_cmd = "sudo yum update"
+        packages_check_cmd = "yum list updates | grep \^fledge"
+    else:
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:133]: I suggest keep as default and avoid extra if

File: python/fledge/services/core/api/update.py [hunk 3]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+    installed_packages_process = await asyncio.create_subprocess_shell(packages_check_cmd,
+                                                                       stdout=asyncio.subprocess.PIPE,
+                                                                       stderr=asyncio.subprocess.PIPE)
+
+    stdout, stderr = await installed_packages_process.communicate()
+    if installed_packages_process.returncode != 0:
+        _logger.info("Nothing to upgrade at the moment. {}".format(stderr))
+        return web.json_response({'updates': []})
+
+    try:
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:155]: add  upgradable_packages = [] before try

File: python/fledge/services/core/api/update.py [hunk 4]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+
+        # remove '' from the list
+        word_list = [w for w in word_list if w != '']
+        packages = []
+
+        # Now match the character / . The string before / is the actual package name we want.
+        for word in word_list:
+            # TODO find regex for yum as well.
+            word_match = re.findall(r".*[/]", word)
+            if len(word_match) > 0:
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:169]: or simply `if len(word_match)`

File: python/fledge/services/core/api/update.py [hunk 5]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+                packages.append(word_match[0].replace('/', '').strip())
+
+        # Make a set to avoid duplicates.
+        upgradable_packages = list(set(packages))
+        return web.json_response({'updates': upgradable_packages})
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:174]: move to else for except

File: python/fledge/services/core/api/update.py [hunk 6]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+    except Exception as ex:
+        msg = "No upgradable packages are found due to {}".format(str(ex))
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:176]: Failed to fetch upgradable packages list for the configured repository! {}

File: python/fledge/services/core/api/update.py [hunk 7]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+    _, stderr = await update_process.communicate()
+    if update_process.returncode != 0:
+        msg = "Could not run {} due to {}".format(update_cmd, stderr.decode('utf-8'))
+        raise web.HTTPInternalServerError(reason=msg, body=json.dumps({"message": msg}))
+
+    installed_packages_process = await asyncio.create_subprocess_shell(packages_check_cmd,
+                                                                       stdout=asyncio.subprocess.PIPE,
+                                                                       stderr=asyncio.subprocess.PIPE)
+
+    stdout, stderr = await installed_packages_process.communicate()
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:150]: stdout as _ or we want to log as INFO?

File: python/fledge/services/core/api/update.py [hunk 8]
```diff
@@ -107,3 +110,69 @@ async def update_package(request):
...
+    if installed_packages_process.returncode != 0:
+        _logger.info("Nothing to upgrade at the moment. {}".format(stderr))
```
Comment (praveen-garg) [inline python/fledge/services/core/api/update.py:152]: We should add a comment why are we considering != 0 exit code as INFO.

Probably we should put stderr in a new log statement after `Nothing to upgrade at the moment, for the configured repository.`

File: python/fledge/services/core/api/update.py (hunk 4 above)
Comment (deepanshu-yadav) [reply python/fledge/services/core/api/update.py:174]: Yes but minor thing.

File: python/fledge/services/core/api/update.py (hunk 7 above)
Comment (deepanshu-yadav) [reply python/fledge/services/core/api/update.py]: we should log in INFO.

**Review by praveen-garg — APPROVED** [review]

**Pull request merged**